File: pythonCodes/credit.py
def main():
    num = int(input("Card number: "))

    strN = str(num)
    if len(strN) != 13 and len(strN) != 15 and len(strN) != 16:
        print("Invalid")
        exit()

    if len(strN) == 16:
        x = 0
    elif len(strN) == 15 or len(strN) == 13:
        x = 1

    valid = credito(x, len(strN), strN)

    if(valid == 1):
        print("Invalid")
    else:
        if(int(strN[0]) == 4):
            print("Visa")
        elif(int(strN[0]) == 3 and (int(strN[1]) == 4 or int(strN[1]) == 7)):
            print("American Express")
        elif(int(strN[0]) == 5 and (int(strN[1]) == 1 or int(strN[1]) == 2 or int(strN[1]) == 3 or int(strN[1]) == 4 or int(strN[1]) == 5)):
            print("Mastercard")

def credito(a, b, strN):
    aux = 0
    for i in range(a, b, 2):
        if int(strN[i])*2 > 9:
            strN1 = str(int(strN[i])*2)
            aux += int(strN1[0]) + int(strN1[1])
        else:
            aux += int(strN[i])*2

    for j in range(abs(a-1), b, 2):
        aux += int(strN[j])

    if aux % 10 == 0:
        return 0
    else:
        return 1


main()

File: cCodes/credit.c
#include <stdio.h>
#include <math.h>

int valid(long int n){
    int sum = 0;
    int k;
    int x;
    int upper;

    if((n/100000000000) < 100){
        //o numero tem 13 digitos
        upper = 13;
    } else if((n/10000000000000) < 100){
        //o numero tem 15 digitos
        upper = 15;
    } else if((n/100000000000000) < 100){
        //o numero tem 16 digitos
        upper = 17;
    } else{
        return 0;
    }

    for(int i = 1; i < upper; i = i + 2){
        k = (n/(long int)pow(10,i))%10;
        x = k*2;
        if(x > 9){
            sum += x/10; //primeiro digito
            sum += x%10; //segundo digito
        } else{
            sum += x;
        }
    }

    for(int i = 0; i < upper; i = i + 2){
        k = (n/(long int)pow(10,i))%10;
        sum += k;
    }

    if(sum % 10 != 0){
        return 0;
    } else{
        if(upper == 15){
            return 1;
        } else if((n/(long int)pow(10,15))%10 == 5){
            return 2;
        } else if((n/(long int)pow(10,15))%10 == 4 || (n/(long int)pow(10,12))%10 == 4){
            return 3;
        } else{
            return 0;
        }
    }
}

int main(void){

    long int n;
    printf("Number: ");
    scanf("%ld", &n);

    if(n/10000000000 < 100){
        while(1){
            printf("Number: ");
            scanf("%ld", &n);
            if(n/10000000000 > 100){
                break;
            }
        }
    }

    if(valid(n) == 0){
        printf("Invalid\n");
    } else if(valid(n) == 1){
        printf("American Express\n");
    } else if(valid(n) == 2){
        printf("MasterCard\n");
    } else{
        printf("Visa\n");
    }
}
